cmd/zdb/zdb.c

```diff
@@ -2202,6 +2202,90 @@ blkid2offset(const dnode_phys_t *dnp, const blkptr_t *bp,
 	    dnp->dn_datablkszsec << SPA_MINBLOCKSHIFT);
 }
 
+static uint32_t
+zfs_get_hdrversion(const zfs_zstdhdr_t *blob)
+{
+	/*
+	 * So we're searching 4 bytes to figure out where the version
+	 * and level bytes are. The level bytes can cover the whole range
+	 * of uint8_t, and so are not helpful. Fortunately, the version
+	 * field is going to have a leading 00 from now until version
+	 * 6.55.56 or higher with how it's represented, so we can dig
+	 * that out, and know that wherever we found it, the two bytes
+	 * "next" to it in the range are the other version bytes, in order,
+	 * and whichever byte remains is the level field.
+	 */
+	uint32_t version = blob->raw_version_level;
+	uint8_t findme = 0xff;
+	int shift;
+	for (shift = 0; shift < 4; shift++) {
+		findme = BF32_GET(version, 8*shift, 8);
+		if (findme == 0)
+			break;
+	}
+	switch (shift) {
+		case 0:
+		version = BSWAP_32(version);
+		version = BF32_GET(version, 8, 24);
+		break;
```

Comment:
From reverse-engineering this code, I think that overall this is decoding a 32-bit number  that can look like one of:

`0x L0 V0 V1 00`
`0x V0 V1 00 L0`
`0x L0 00 V1 V0`
`0x 00 V1 V0 L0`
where 00 is a literal 0, L0 is the level, V0 is the low bits of the version and V1 is the high bits of the version.

I would appreciate seeing this explanation in the comments, as well as an explanation of how the on-disk format could come to be this way.

Reply:
Ah, I'm terribly sorry, I thought your prior request was explicitly "an explanation for why we can't use standard macros for this", which I thought was alleviated by doing so.

I have, to be clear, only seen two of these four cases in the wild (2 for LE and 3 for BE), outside of bytes I mangled with my own code modifications; I just included all the cases (and tested them, with aforementioned manually-induced-mangling) for completeness.

I'll include an explanation for the aforementioned two, but it will amount to "we wrote a representation which is not guaranteed to be consistent across platforms to disk".

Reply:
Yeah I think we could be more specific, e.g. saying that it used a bitfield and the order of the elements of the bitfield are not specified by the C language, so it might be version-first or level-first.  Although in practice you could check if any of our supported platforms actually did it in an other-than expected order, which might simplify this code.

Reply:
I've checked Linux/{aarch64,mips64el,ppc64,riscv64,sparc64,x86_64,x86} and FreeBSD/{x86_64,ppc64}, and the only difference is that BE did one case and LE did another. So I could hardcode checks for the two cases instead of all 4, but it doesn't feel like it simplifies much, since it's still going to require a nontrivial explanation for the two different bit configurations.

Reply:
There, I replaced the existing comment with one with diagrams. Please let me know if it's not clear enough.